### PR TITLE
Adding custom serialization for Request object.

### PR DESCRIPTION
**Pull request by safaci2000**

### CODE REVIEW

**Review by safaci2000 — COMMENTED** [review]
Basically I needed a way to pass the Folder ID, this is my attempt to do so.  Please let me know what you think of the Impl.  The tests should qualify the Implementation and ensure everything still works.

File: rest-dashboard_test.go [hunk 1]
```diff
@@ -283,3 +285,72 @@ func testUIntSearchParam(t *testing.T, sp func(uint) sdk.SearchParam, key string
...
+	r := getRawBoardRequest(true)
+	//Serialize object
+	data, err := json.Marshal(r)
+	if err != nil {
+		panic(err)
+	}
+	//Deserialize data
+	var rawData map[string]interface{} = make(map[string]interface{})
+	err = json.Unmarshal(data, &rawData)
+	if err != nil {
```
Comment (GiedriusS) [inline rest-dashboard_test.go:299]: Maybe `assert.Nil(t, json.Unmarshal(data, &rawData))`?

File: rest-dashboard_test.go [hunk 2]
```diff
@@ -283,3 +285,72 @@ func testUIntSearchParam(t *testing.T, sp func(uint) sdk.SearchParam, key string
 		}
 	}
 }
+
+func Test_DecodeRawBoardPreserveId(t *testing.T) {
+	r := getRawBoardRequest(true)
+	//Serialize object
+	data, err := json.Marshal(r)
```
Comment (GiedriusS) [inline rest-dashboard_test.go:292]: Same here.

File: rest-dashboard.go [hunk 3]
```diff
@@ -271,31 +312,20 @@ func (r *Client) SetDashboard(ctx context.Context, board Board, params SetDashbo
...
 		err     error
-		buf     bytes.Buffer
-		plain   = make(map[string]interface{})
 	)
-	if err = json.Unmarshal(raw, &plain); err != nil {
-		return StatusMessage{}, err
+	raw, err := json.Marshal(request)
+
+	if err != nil {
+		panic(err)
```
Comment (GiedriusS) [inline rest-dashboard.go:326]: Don't ever, ever `panic()` in a library. We don't want to crash our users' programs. Please return `StatusMessage{}, err`

**Review by GiedriusS — CHANGES_REQUESTED** [review]

**Comment by GiedriusS** [comment]
Thanks for starting to use `testify`, btw, I wanted to start using some kind of library for those checks to reduce the verbosity a little bit

File: rest-dashboard_test.go (hunk 1 above)
Comment (safaci2000) [reply rest-dashboard_test.go:299]: Done.

File: rest-dashboard.go (hunk 3 above)
Comment (safaci2000) [reply rest-dashboard.go:326]: Good point, Done

**Comment by safaci2000** [comment]
I'm not sure if testify is the best library to use but I kept on running into it in various blogs.  

One question, I haven't been able to get all the tests to pass locally.  I imagine some of them require an instance of grafana to be up and running.  I can throw a quick docker-compose in a future commit to support testing.  Is there any docs on running the tests locally I missed?

File: rest-dashboard_test.go (hunk 2 above)
Comment (safaci2000) [reply rest-dashboard_test.go:292]: Done

**Comment by GiedriusS** [comment]
> I'm not sure if testify is the best library to use but I kept on running into it in various blogs.
> 
> One question, I haven't been able to get all the tests to pass locally. I imagine some of them require an instance of grafana to be up and running. I can throw a quick docker-compose in a future commit to support testing. Is there any docs on running the tests locally I missed?

Yep, you need a local Grafana for the integration tests. Typically I just run these beforehand: https://github.com/grafana-tools/sdk/blob/master/.travis.yml#L96-L99 but yes, it would be nice to have a docker-compose file

**Comment by safaci2000** [comment]
Thanks for the update @GiedriusS.  Did you have any other comments in this PR  you'd like me to address?

**Comment by GiedriusS** [comment]
Sorry for the delay, I'll take a look in a bit

**Comment by GiedriusS** [comment]
@safaci2000 could you please rebase on the newest `master`? It has GH actions and everything. It would be great to have a green tick before merging this :+1:

**Comment by safaci2000** [comment]
@GiedriusS Done.  Rebased.  Ready for your review!

**Comment by safaci2000** [comment]
@GiedriusS  friendly poke when you have a chance to look at this again!  Thanks!

File: rest-dashboard.go [hunk 4]
```diff
@@ -57,6 +57,46 @@ type BoardProperties struct {
...
+		var plain map[string]interface{} = make(map[string]interface{})
+		if err := json.Unmarshal(d.Dashboard, &plain); err != nil {
+			return nil, err
+		}
+		plain["id"] = 0
+		raw, _ = json.Marshal(plain)
+
+	}
+
+	parameters, err := json.Marshal(d.Parameters)
```
Comment (GiedriusS) [inline rest-dashboard.go:84]: The code below seems a bit hacky. Why not:
```
  var output struct {
    Dashboard []byte
    SetDashboardParams
  }
  
  output.Dashboard = raw
  output.SetDashboardParams = parameters
  
  return json.Marshal(output)
```
:thinking: ? It's much simpler and should do the same. (https://play.golang.org/p/nssZ3HPJOjG)

File: rest-dashboard.go [hunk 5]
```diff
@@ -57,6 +57,46 @@ type BoardProperties struct {
...
+	buf.WriteString(`{"dashboard":`)
+	buf.Write(raw)
+	buf.WriteString(",")
+	//Remove wrapping {}
+	parametersData := string(parameters)[1 : len(parameters)-1]
+	buf.Write([]byte(parametersData))
+	buf.WriteString("}")
+
+	return buf.Bytes(), nil
+
```
Comment (GiedriusS) [inline rest-dashboard.go:97]: Spurious new line?

File: rest-dashboard_test.go [hunk 6]
```diff
@@ -283,3 +285,61 @@ func testUIntSearchParam(t *testing.T, sp func(uint) sdk.SearchParam, key string
...
+	err = json.Unmarshal(data, &rawData)
+	assert.Nil(t, err)
+	folderId := rawData["FolderID"]
+
+	assert.Equal(t, folderId.(float64), float64(27))
+	assert.Equal(t, rawData["Overwrite"].(bool), true)
+	rawData = rawData["dashboard"].(map[string]interface{})
+	assert.Equal(t, rawData["id"].(float64), float64(0))
+	assert.Equal(t, rawData["title"].(string), "woot")
+
```
Comment (GiedriusS) [inline rest-dashboard_test.go:324]: Spurious new line.

File: rest-dashboard_test.go [hunk 7]
```diff
@@ -283,3 +285,61 @@ func testUIntSearchParam(t *testing.T, sp func(uint) sdk.SearchParam, key string
...
+	assert.Equal(t, folderId.(float64), float64(27))
+	assert.Equal(t, rawData["Overwrite"].(bool), true)
+	rawData = rawData["dashboard"].(map[string]interface{})
+	assert.Equal(t, rawData["id"].(float64), float64(25))
+	assert.Equal(t, rawData["title"].(string), "woot")
+
+}
+
+func Test_DecodeRawBoardWipeId(t *testing.T) {
+
```
Comment (GiedriusS) [inline rest-dashboard_test.go:308]: Spurious new line.

File: rest-dashboard_test.go [hunk 8]
```diff
@@ -283,3 +285,61 @@ func testUIntSearchParam(t *testing.T, sp func(uint) sdk.SearchParam, key string
...
+	params := sdk.SetDashboardParams{
+		FolderID:   27,
+		Overwrite:  true,
+		PreserveId: preserveId,
+	}
+	r.Dashboard = rawBoard
+	r.Parameters = params
+
+	return r
+
```
Comment (GiedriusS) [inline rest-dashboard_test.go:344]: Spurious line.

File: rest-dashboard.go [hunk 9]
```diff
@@ -219,11 +259,12 @@ func (r *Client) Search(ctx context.Context, params ...SearchParam) ([]FoundBoar
...
 
-// SetDashboardParams contains the extra parameteres
+// SetDashboardParams contains the extra parameters
 // that affects where and how the dashboard will be stored
 type SetDashboardParams struct {
-	FolderID  int
-	Overwrite bool
+	FolderID   int
+	Overwrite  bool
+	PreserveId bool `json:"-"` //ignore on serialization
```
Comment (GiedriusS) [inline rest-dashboard.go:267]:
```suggestion
	PreserveId bool `json:"-"`
```

**Review by GiedriusS — COMMENTED** [review]
Just a few suggestions on how I think this could be made a bit simpler but overall LGTM :muscle:

File: rest-dashboard.go (hunk 4 above)
Comment (safaci2000) [reply rest-dashboard.go:84]: yeah that's way cleaner.  I was trying to figure out a way to do that without string manipulation.  Updated.

File: rest-dashboard.go (hunk 5 above)
Comment (safaci2000) [reply rest-dashboard.go:97]: Done

File: rest-dashboard_test.go (hunk 6 above)
Comment (safaci2000) [reply rest-dashboard_test.go:324]: Done

File: rest-dashboard_test.go (hunk 7 above)
Comment (safaci2000) [reply rest-dashboard_test.go:308]: Done

File: rest-dashboard_test.go (hunk 8 above)
Comment (safaci2000) [reply rest-dashboard_test.go]: Done

File: rest-dashboard.go (hunk 9 above)
Comment (safaci2000) [reply rest-dashboard.go:267]: Applied

File: rest-dashboard.go [hunk 10]
```diff
@@ -57,6 +57,46 @@ type BoardProperties struct {
...
+//RawBoardRequest struct that wraps Board and parameters being sent
+type RawBoardRequest struct {
+	Dashboard  []byte
+	Parameters SetDashboardParams
+}
+
+//MarshalJSON serializes the request to match the expectations of the grafana API.
+// Additionally, if preseveID is false, then the dashboard id is set to 0
+func (d RawBoardRequest) MarshalJSON() ([]byte, error) {
+	buf := bytes.NewBufferString("")
```
Comment (GiedriusS) [inline rest-dashboard.go:69]: This seems unused :smile:

File: rest-dashboard.go (hunk 4 above)
Comment (safaci2000) [reply rest-dashboard.go]: This is now resolved!

File: rest-dashboard.go (hunk 10 above)
Comment (safaci2000) [reply rest-dashboard.go]: Done.

**Review by GiedriusS — APPROVED** [review]
TBH I'm not a huge fan of these raw methods that accept a []byte because they seem more like a workaround for `sdk.Board` not having all of the fields. But, for now, I don't think they should be removed so let's add this method. Thank you :muscle:

**Comment by safaci2000** [comment]
@GiedriusS Ideally I wouldn't use any of the Raw methods but I've had it fail hard when I tried not to use the Raw calls.  Thanks for accepting the PR.

**Pull request merged**